File: python/bigasterisk/fuzz.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Fuzz testing for Spark SQL: generate inputs for a query and see what breaks.

Testing a data-intensive application normally means running it over a real dataset,
which is slow and covers only the cases that dataset happens to contain. Fuzzing
generates small inputs instead, and steers them toward parts of the query nothing has
exercised yet.

Three mutation strategies, which are the three fuzzers this platform unifies:

``random``
    Values drawn at random for each column's type. Cheap, good at malformed values,
    poor at getting past a join. (BigFuzz)
``natural``
    Values spliced column-wise out of rows already seen, so generated rows look like
    real data while the combinations are new. (NaturalFuzz)
``co-dependent``
    Splicing that draws joined columns from a shared pool, so generated rows survive
    the join. The default. (DepFuzz)

    import bigasterisk

    result = bigasterisk.fuzz(spark).fuzz(
        "SELECT c.name, SUM(o.amount) FROM orders o JOIN customers c ON o.cid = c.cid "
        "GROUP BY c.name",
        {"orders": orders, "customers": customers},
        iterations=100)

    print(result.coverage)
    for failure in result.failures:
        print(failure)
"""

import json

from .query import as_query


class FuzzFailure:
    """An input that made the query fail."""

    def __init__(self, payload):
        self.iteration = payload["iteration"]
        self.error = payload["error"]
        self.tables = payload["tables"]

    def __repr__(self):
        rows = "\n".join(
            "  %s: %s" % (name, ", ".join(rows)) for name, rows in self.tables.items())
        return "iteration %d: %s\n%s" % (self.iteration, self.error, rows)


class FuzzResult:
    """What a fuzzing campaign found."""

    def __init__(self, payload):
        self.iterations = payload["iterations"]
        self.total_branches = payload["totalBranches"]
        self.empty_results = payload["emptyResults"]
        self.coverage = payload["coverage"]
        self.abstracted = payload["abstracted"]
        self.covered = set(payload["covered"])
        self.failures = [FuzzFailure(f) for f in payload["failures"]]
        #: a few of the generated inputs, so you can see what the campaign fed the query
        self.samples = [FuzzSample(s) for s in payload.get("samples", [])]

    def __repr__(self):
        return ("FuzzResult(%d iterations (%d without Spark), %d failures, "
                "coverage %.0f%% of %d branches, %d empty)" % (
                    self.iterations, self.abstracted, len(self.failures),
                    self.coverage * 100, self.total_branches, self.empty_results))


class FuzzSample:
    """One input the campaign generated, kept so it can be looked at."""

    def __init__(self, payload):
        self.iteration = payload["iteration"]
        #: whether it reached a branch nothing had reached before
        self.reached_new = payload["reachedNew"]
        #: whether the query returned nothing for it
        self.empty = payload["empty"]
        #: the generated rows, by table name, each row rendered as text
        self.tables = payload["tables"]

    def __repr__(self):
        note = []
        if self.reached_new:
            note.append("reached new coverage")
        if self.empty:
            note.append("produced no output")
        suffix = "  (%s)" % ", ".join(note) if note else ""
        body = "\n".join(
            "  %s: %s%s" % (name, ", ".join(rows[:4]),
                            ", ... %d rows" % len(rows) if len(rows) > 4 else "")
            for name, rows in sorted(self.tables.items()))
        return "iteration %d%s\n%s" % (self.iteration, suffix, body)


class Fuzz:
    """Fuzz testing for a SparkSession."""

    def __init__(self, spark):
        self._spark = spark
        self._support = spark._jvm.org.bigasterisk.api.BigAsterisk.fuzz(
            spark._jsparkSession)

    def fuzz(self, query, seeds, iterations=100, strategy="co-dependent",
             rows_per_table=10, seed=0, guided=True, abstract_framework=True,
             rows_per_vector=3, keep_samples=3):
        """Run a fuzzing campaign against ``query``.

        ``query`` is a DataFrame — the pipeline itself — or a SQL string. Either way
        the campaign has to run it repeatedly with data of its own choosing; for a
        DataFrame that is done by substituting into its plan, so each seed must be the
        DataFrame the pipeline was built from (or a table it reads under that name).

        ``seeds`` maps each table name the query reads to a DataFrame. Their rows are
        the corpus generated values are drawn from; only the schema matters for the
        ``random`` strategy.

        With ``abstract_framework`` (the default), iterations are evaluated by
        interpreting the query's plan over in-memory rows rather than running a Spark
        job — the operator semantics are Spark's either way, but the planning,
        scheduling and task setup that dwarf a twenty-row query are gone. Anything the
        interpreter does not support falls back to Spark, so this changes speed and
        never results.

        ``keep_samples`` retains that many of the generated inputs on the result, so you
        can look at what the campaign actually fed the query — see ``FuzzResult.samples``.
        Inputs that reached new coverage are preferred over merely early ones.

        ``rows_per_vector`` bounds the corpus: seed rows that decide every branch of the
        query the same way are interchangeable to its control flow, so only a few of each
        distinct behaviour are kept.

        The campaign swaps generated data in under those table names while it runs and
        restores the originals afterwards.
        """
        jseeds = self._spark._jvm.java.util.HashMap()
        for name, df in seeds.items():
            jseeds.put(name, df._jdf)
        result = self._support.fuzzJava(
            as_query(query), jseeds, int(iterations), strategy, int(rows_per_table),
            int(seed), bool(guided), bool(abstract_framework), int(rows_per_vector),
            int(keep_samples))
        return FuzzResult(json.loads(result.json()))

File: modules/api/src/main/scala/org/bigasterisk/api/Fuzz.scala
package org.bigasterisk.api

import org.apache.spark.sql.{DataFrame, Row}

/**
 * How new test inputs are derived from the ones already seen.
 *
 * The three strategies are the three fuzzers this platform unifies. They differ only in
 * how a candidate row is built, which is why they share one loop rather than three.
 *
 * @group fuzz
 */
sealed trait MutationStrategy { def name: String }

object MutationStrategy {

  /**
   * Values drawn at random for each column's type.
   *
   * The baseline: cheap, and good at finding crashes on malformed values. Poor at
   * getting past a join, since a randomly generated key almost never matches.
   * Corresponds to BigFuzz.
   */
  case object Random extends MutationStrategy { val name = "random" }

  /**
   * Values spliced column-wise out of rows already seen.
   *
   * Every value is one that genuinely occurred in that column, so generated rows look
   * like real data — the right shape, the right formats, plausible magnitudes — while
   * the *combinations* are new. Corresponds to NaturalFuzz.
   */
  case object Natural extends MutationStrategy { val name = "natural" }

  /**
   * Splicing that respects co-dependence between tables.
   *
   * A join makes two columns of two tables dependent: mutate one freely and the rows
   * stop matching, so the query returns nothing and the fuzzer learns nothing. This
   * draws joined columns from the same pool on both sides, so generated rows survive
   * the join. Corresponds to DepFuzz.
   */
  case object CoDependent extends MutationStrategy { val name = "co-dependent" }

  val all: Seq[MutationStrategy] = Seq(Random, Natural, CoDependent)

  /** Looks a strategy up by [[MutationStrategy.name]]. */
  def byName(name: String): MutationStrategy =
    all.find(_.name.equalsIgnoreCase(name)).getOrElse {
      throw new IllegalArgumentException(
        s"unknown mutation strategy '$name'; expected one of ${all.map(_.name).mkString(", ")}")
    }
}

/**
 * How a fuzzing campaign is run.
 *
 * @param iterations   how many candidate inputs to try
 * @param strategy     how candidates are derived
 * @param rowsPerTable rows to generate for each table
 * @param seed         random seed, so a campaign reproduces exactly
 * @param guided       when true, a candidate that reaches a branch nothing had reached
 *                     is kept and mutated further; when false every candidate is drawn
 *                     from the seed data alone
 * @param abstractFramework when true, iterations are evaluated by interpreting the
 *                     query's plan over in-memory rows instead of running a Spark job.
 *                     The operator semantics are Spark's either way; what goes is the
 *                     planning, scheduling and task setup that dwarfs a twenty-row query.
 *                     Any query the interpreter does not support falls back to Spark, so
 *                     this changes speed and never results.
 *
 * @group fuzz
 */
case class FuzzConfig(
    iterations: Int = 100,
    strategy: MutationStrategy = MutationStrategy.CoDependent,
    rowsPerTable: Int = 10,
    seed: Long = 0L,
    guided: Boolean = true,
    abstractFramework: Boolean = true) {
  require(iterations >= 0, s"iterations must not be negative, got $iterations")
  require(rowsPerTable > 0, s"rowsPerTable must be positive, got $rowsPerTable")
}

/**
 * An input that made the query fail.
 *
 * @param iteration when it was found
 * @param error     the exception's type and message
 * @param tables    the generated rows, by table name — enough to reproduce the failure
 *
 * @group fuzz
 */
case class FuzzFailure(iteration: Int, error: String, tables: Map[String, Seq[Row]]) {
  override def toString: String =
    s"iteration $iteration: $error\n" +
      tables.map { case (t, rows) => s"  $t: ${rows.mkString(", ")}" }.mkString("\n")
}

/**
 * What a fuzzing campaign found.
 *
 * @param iterations      candidates actually run
 * @param failures        inputs that made the query fail
 * @param covered         branches of the query that some input reached
 * @param totalBranches   branches the query has
 * @param emptyResults    candidates that produced no output rows at all — the symptom
 *                        of inputs that cannot get past a join
 * @param abstracted      iterations evaluated without Spark. The rest fell back, either
 *                        because the abstraction was switched off or because the query
 *                        left the interpreter's supported set.
 *
 * @group fuzz
 */
case class FuzzResult(
    iterations: Int,
    failures: Seq[FuzzFailure],
    covered: Set[String],
    totalBranches: Int,
    emptyResults: Int,
    abstracted: Int = 0) {

  /** Fraction of iterations that avoided Spark entirely. */
  def abstractionRatio: Double =
    if (iterations == 0) 0.0 else abstracted.toDouble / iterations

  /** Fraction of the query's branches that some generated input reached. */
  def coverage: Double = if (totalBranches == 0) 1.0 else covered.size.toDouble / totalBranches

  override def toString: String =
    f"FuzzResult($iterations iterations ($abstracted without Spark), " +
      f"${failures.size} failures, coverage ${coverage * 100}%.0f%% of $totalBranches " +
      f"branches, $emptyResults empty)"

  /**
   * The result as a JSON object.
   *
   * Provided for language bindings that cannot marshal Scala collections across the
   * process boundary — the PySpark front end reads this. Generated rows are rendered
   * with their `toString`, which is enough to reproduce a failure by eye.
   */
  def json: String = {
    def quote(s: String): String =
      "\"" + s.flatMap {
        case '"'              => "\\\""
        case '\\'             => "\\\\"
        case '\n'             => "\\n"
        case '\r'             => "\\r"
        case '\t'             => "\\t"
        case c if c.isControl => f"\\u${c.toInt}%04x"
        case c                => c.toString
      } + "\""

    val failuresJson = failures.map { f =>
      val tables = f.tables.map { case (name, rows) =>
        s"${quote(name)}:[${rows.map(r => quote(r.toString)).mkString(",")}]"
      }.mkString(",")
      s"""{"iteration":${f.iteration},"error":${quote(f.error)},"tables":{$tables}}"""
    }.mkString(",")

    s"""{"iterations":$iterations,"totalBranches":$totalBranches,""" +
      s""""emptyResults":$emptyResults,"coverage":$coverage,"abstracted":$abstracted,""" +
      s""""covered":[${covered.toSeq.sorted.map(quote).mkString(",")}],""" +
      s""""failures":[$failuresJson]}"""
  }
}

/**
 * Fuzz testing for Spark SQL: generate inputs for a query and see what breaks.
 *
 * Testing a data-intensive application normally means running it over a real dataset,
 * which is slow and covers only the cases that dataset happens to contain. Fuzzing
 * generates small inputs instead, and steers them toward parts of the query nothing has
 * exercised yet.
 *
 * Obtain an instance from [[BigAsterisk.fuzz]].
 *
 * {{{
 * val result = BigAsterisk.fuzz(spark).fuzz(
 *   "SELECT c.name, SUM(o.amount) FROM orders o JOIN customers c ON o.cid = c.cid " +
 *   "GROUP BY c.name",
 *   Map("orders" -> orders, "customers" -> customers))
 *
 * result.coverage          // how much of the query the campaign reached
 * result.failures.foreach(println)
 * }}}
 *
 * @group fuzz
 */
trait FuzzSupport {

  /**
   * Runs a fuzzing campaign against `query`.
   *
   * @param seeds the tables the query reads, by the name it reads them under. Their
   *        rows are the corpus generated values are drawn from; only the schema is
   *        required for [[MutationStrategy.Random]].
   */
  def fuzz(query: String, seeds: Map[String, DataFrame], config: FuzzConfig): FuzzResult

  /** Runs a campaign with the default configuration. */
  final def fuzz(query: String, seeds: Map[String, DataFrame]): FuzzResult =
    fuzz(query, seeds, FuzzConfig())

  /**
   * [[fuzz]] for callers that cannot build a Scala `Map` or a [[FuzzConfig]] — Java,
   * and Py4J, which marshals a Python dict to `java.util.Map`.
   */
  final def fuzzJava(
      query: String,
      seeds: java.util.Map[String, DataFrame],
      iterations: Int,
      strategy: String,
      rowsPerTable: Int,
      seed: Long,
      guided: Boolean,
      abstractFramework: Boolean): FuzzResult = {
    import scala.jdk.CollectionConverters._
    fuzz(
      query,
      seeds.asScala.toMap,
      FuzzConfig(iterations, MutationStrategy.byName(strategy), rowsPerTable, seed, guided,
        abstractFramework))
  }
}
